### PR TITLE
Cleanup more GCC warnings

### PR DESCRIPTION
**Pull request by franksinankaya**
Suppress will never be NULL warnings
Convert multi-character literal
Fix mutli-line comment warning
Remove hr=hr undefined assignment
Move declaration into same file as one was defined Extern the other one was static
Add parenthesis

### CODE REVIEW

**Comment by franksinankaya** [comment]
@jkotas @janvorli @am11

File: src/vm/CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,9 @@ include_directories(${ARCH_SOURCES_DIR})
 add_definitions(-DUNICODE)
 add_definitions(-D_UNICODE)
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+add_compile_options(-Wno-address)
+endif()
```
Comment (am11) [inline src/vm/CMakeLists.txt]: Can we fix the source instead? See https://dmalcolm.fedorapeople.org/gcc/2015-08-31/rst-experiment/options-to-request-or-suppress-warnings.html#cmdoption-Waddress

Comment (franksinankaya) [reply src/vm/CMakeLists.txt]: This is the warning. Let me know:

In file included from src/inc/stgpool.h:28:0,
                 from src/inc/../md/inc/metamodel.h:18,
                 from src/inc/../md/inc/metamodelro.h:19,
                 from src/inc/metadata.h:17,
                 from src/vm/util.hpp:19,
                 from src/vm/common.h:110,
                 from src/vm/commodule.cpp:6:
src/vm/commodule.cpp: In function ‘Object* GetTypesInner(Module*)’:
src/vm/commodule.cpp:1066:29: warning: the address of ‘throwable’ will never be NULL [-Waddress]
         EX_CATCH_THROWABLE(&throwable);
		 
		 
src/vm/invokeutil.cpp: In static member function ‘static void InvokeUtil::SetValidField(CorElementType, TypeHandle, FieldDesc*, OBJECTREF*, OBJECTREF*, TypeHandle, CLR_BOOL*)’:
src/vm/invokeutil.cpp:978:29: warning: the address of ‘Throwable’ will never be NULL [-Waddress]
         EX_CATCH_THROWABLE(&Throwable);
                             ^
src/inc/ex.h:1087:21: note: in definition of macro ‘EX_CATCH_THROWABLE’
         if (NULL != ppThrowable)                                                \
                     ^
src/vm/invokeutil.cpp: In static member function ‘static OBJECTREF InvokeUtil::GetFieldValue(FieldDesc*, TypeHandle, OBJECTREF*, TypeHandle, CLR_BOOL*)’:
src/vm/invokeutil.cpp:1186:29: warning: the address of ‘Throwable’ will never be NULL [-Waddress]
         EX_CATCH_THROWABLE(&Throwable);
                             ^
src/inc/ex.h:1087:21: note: in definition of macro ‘EX_CATCH_THROWABLE’
         if (NULL != ppThrowable)                                                \

Comment (am11) [reply src/vm/CMakeLists.txt]: Looking at `invokeutil.cpp` and `commodule.cpp`, the *address* of `Throwable` can certainly not be `NULL`; we can simply remove those three lines?

Comment (janvorli) [reply src/vm/CMakeLists.txt]: Yes, please remove the check. We always pass it a reference to a local to this macro and I don't see why we would want to pass it NULL.

File: src/classlibnative/bcltype/varargsnative.cpp [hunk 2]
```diff
@@ -87,8 +87,8 @@ static void InitCommon(VARARGS *data, VASigCookie** cookie)
     //  STACK_GROWS_DOWN_ON_ARGS_WALK
 
     //   <return address>  ;; lower memory                  
-    //   <varargs_cookie>         \
-    //   <argN>                    \
+    //   <varargs_cookie>         '\'
```
Comment (janvorli) [inline src/classlibnative/bcltype/varargsnative.cpp]: Hmm, I've tried this in the compiler explorer and it didn't show any warnings for GCC (arm / arm64 / x64) tried version 4.5.4 and 8.2.

File: src/ToolBox/SOS/Strike/sildasm.cpp [hunk 3]
```diff
@@ -65,6 +65,7 @@ static OpCode opcodes[] =
 
 static ULONG position = 0;
 static BYTE *pBuffer = NULL;
+static char* asString(CQuickBytes *out);
```
Comment (janvorli) [inline src/ToolBox/SOS/Strike/sildasm.cpp]: Why was this move needed?

Comment (franksinankaya) [reply src/ToolBox/SOS/Strike/sildasm.cpp]: It is defined an extern function in the header file today yet the definition in the source file is static. GCC warns as signature failure. I looked to see if the function is shared with other files. The answer was no. I ended up moving the definition inside the source file where it is needed.

File: src/classlibnative/bcltype/varargsnative.cpp (hunk 2 above)
Comment (franksinankaya) [reply src/classlibnative/bcltype/varargsnative.cpp]: I get multiline comment warning with 

Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.11)

using ubuntu 16.04.

File: src/vm/CMakeLists.txt (hunk 1 above)
Comment (franksinankaya) [reply src/vm/CMakeLists.txt]: done

File: src/classlibnative/bcltype/varargsnative.cpp (hunk 2 above)
Comment (janvorli) [reply src/classlibnative/bcltype/varargsnative.cpp]: Hmm, it should not show the warning unless we compile with `-Wall` or `-Wcomment`. Strange.

Comment (franksinankaya) [reply src/classlibnative/bcltype/varargsnative.cpp]: We are compiling with -Wall. This file is a copy/paste of clang file.

https://github.com/dotnet/coreclr/blob/439193cf9e63678138a74d8020dd98a7096c5d49/src/pal/tools/gcc-compiler-override.txt#L9

**Comment by franksinankaya** [comment]
@dotnet-bot test  Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test

**Comment by franksinankaya** [comment]
@dotnet-bot retest Windows_NT x64 Checked CoreFX Tests

**Comment by franksinankaya** [comment]
@dotnet-bot retest Windows_NT x64 Checked CoreFX Tests

**Comment by franksinankaya** [comment]
rebased to master

**Comment by franksinankaya** [comment]
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked crossgen_comparison Build and Test

**Comment by franksinankaya** [comment]
is this good to go in too?

File: src/zap/zapimage.cpp [hunk 4]
```diff
@@ -1025,8 +1025,8 @@ HANDLE ZapImage::GenerateFile(LPCWSTR wszOutputFileName, CORCOMPILE_NGEN_SIGNATU
     {    
         // Write the debug directory entry for the NGEN PDB
         RSDS rsds = {0};
-        
-        rsds.magic = 'SDSR';
+
+        rsds.magic = (DWORD)('S' * 0x1000000 +  'D'*0x10000 + 'S' * 0x100 + 'R');
```
Comment (jkotas) [inline src/zap/zapimage.cpp]: Nit: Inconsistent spacing around `*`

Comment (jkotas) [reply src/zap/zapimage.cpp]: This is what the same magic looks in other places: https://github.com/dotnet/coreclr/search?q=0x53445352&unscoped_q=0x53445352 . Maybe nice to use the same pattern.

Comment (franksinankaya) [reply src/zap/zapimage.cpp]: sure

**Comment by franksinankaya** [comment]
@dotnet-bot retest Windows_NT x64 Release CoreFX Tests

**Comment by franksinankaya** [comment]
nice, any other review comments?

**Pull request merged**